### PR TITLE
Share an email thread to workspace members chip and dropdown  (#4199)

### PR DESCRIPTION
**Pull request by pereira0x**
# Feature: Email thread members visibility

For this feature we implemented a chip and a dropdown menu that allows users to check which workspace members can see an email thread, as depicted on issue (#4199). 

## Implementations

- create a new database table (messageThreadMember)
- relations between `messageThreadMembers` and the relevant existing tables (`MessageThread` and `WorkspaceMembers`)
- added a new column to the `MessageThread table`: `everyone` - to indicate that all workspace members can see the email thread
- create a new repository for the new table, including new queries
- edit the queries so that the new fields could be fetched from the frontend
- created a component `MultiChip`, that shows a group of user avatars, instead of just one
- created a component, `ShareDropdownMenu`, that shows up once the `EmailThreadMembersChip` is clicked. On this menu you can see which workspace members can view the email thread.

## Screenshots

Here are some screenshots of the frontend components that were created:

Chip with everyone in the workspace being part of the message thread:
image

Chip with just one member of the workspace (the owner) being part of the message thread:
image

Chip with some members of the workspace being part of the message thread:
image

How the chip looks in a message thread:
image

Dropdown that opens when you click on the chip:
image

## Testing and Mock data

We also added mock data (TypeORM seeds), focusing on adding mock data related to message thread members.

## Conclusion

As some of the changes that we needed to do, regarding the change of visibility of the message thread, were not covered by the existing documentation, we were told to open a PR and ask for feedback on this part of the implementation. Right now, our implementation is focused on displaying who is part of an email thread.

Feel free to let us know which steps we should follow next :)

### CODE REVIEW

**Comment by Bonapara** [comment]
Thanks so much @pereira0x! From a design perspective, there is an unwanted top bar on the following component:

> Dropdown that opens when you click on the chip: image

**Comment by simaosanguinho** [comment]
Hey @Bonapara. Me and @pereira0x will take care of removing the unwanted top bar from the Dropdown component. Additionally, we'll review the bot suggestions and apply them to our patch.

File: packages/twenty-front/src/modules/activities/emails/components/EmailThreadMembersChip.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { MessageThread } from '@/activities/emails/types/MessageThread';
+import { SharedDropdownMenu } from '@/ui/layout/dropdown/components/SharedDropdownMenu';
+
+export const EmailThreadMembersChip = ({
+  messageThread,
+}: {
+  messageThread: MessageThread | null;
+}) => {
+  const renderChip = () => {
+    if (!messageThread) {
```
[the remark this answers is not in the document]
Comment (simaosanguinho) [reply packages/twenty-front/src/modules/activities/emails/components/EmailThreadMembersChip.tsx:10]: We don't feel the need for this change, as we wanted the variable to be consistent with the database table.

File: packages/twenty-front/src/modules/activities/emails/components/EmailThreadMembersChip.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (pereira0x) [reply packages/twenty-front/src/modules/activities/emails/components/EmailThreadMembersChip.tsx:12]: This is correct, we will be fixing this in our next commit.

File: packages/twenty-front/src/modules/activities/emails/right-drawer/components/RightDrawerEmailThread.tsx [hunk 3]
```diff
@@ -10,6 +11,7 @@ import { useRightDrawerEmailThread } from '@/activities/emails/right-drawer/hook
 import { emailThreadIdWhenEmailThreadWasClosedState } from '@/activities/emails/states/lastViewableEmailThreadIdState';
 import { EmailThreadMessage as EmailThreadMessageType } from '@/activities/emails/types/EmailThreadMessage';
 import { RIGHT_DRAWER_CLICK_OUTSIDE_LISTENER_ID } from '@/ui/layout/right-drawer/constants/RightDrawerClickOutsideListener';
+import { messageThreadState } from '@/ui/layout/right-drawer/states/messageThreadState';
```
[the remark this answers is not in the document]
Comment (simaosanguinho) [reply packages/twenty-front/src/modules/activities/emails/right-drawer/components/RightDrawerEmailThread.tsx:14]: We will apply this change as well.

File: packages/twenty-server/src/database/typeorm-seeds/workspace/message-thread-members.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
+import { EntityManager } from 'typeorm';
```
[the remark this answers is not in the document]
Comment (pereira0x) [reply packages/twenty-server/src/database/typeorm-seeds/workspace/message-thread-members.ts:1]: This is not an existing practice in the codebase. We don't think this is relevant here.

File: packages/twenty-server/src/database/typeorm-seeds/workspace/message-thread-members.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  await entityManager
+    .createQueryBuilder()
+    .insert()
+    .into(`${schemaName}.${tableName}`, [
+      'id',
+      'createdAt',
+      'updatedAt',
+      'deletedAt',
+      'messageThreadId',
+      'workspaceMemberId',
```
[the remark this answers is not in the document]
Comment (pereira0x) [reply packages/twenty-server/src/database/typeorm-seeds/workspace/message-thread-members.ts:41]: This is a good recommendation. We will be applying this on our next commit.

File: packages/twenty-server/src/database/typeorm-seeds/workspace/message-thread-members.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    ])
+    .orIgnore()
+    .values([
+      {
```
[the remark this answers is not in the document]
Comment (pereira0x) [reply packages/twenty-server/src/database/typeorm-seeds/workspace/message-thread-members.ts:45]: Same as above. We will be fixing this..

File: packages/twenty-ui/src/display/chip/components/MultiChip.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  maxWidth?: number;
+};
+
+export const MultiChip = ({
+  linkToEntity,
+  names = [],
+  avatarUrls = [],
+  LeftIcon,
+  RightIcon,
+  className,
```
[the remark this answers is not in the document]
Comment (pereira0x) [reply packages/twenty-ui/src/display/chip/components/MultiChip.tsx:36]: We can change this variable name, but we think that a better name would be `numberOfAvatars`.

File: packages/twenty-front/src/modules/ui/layout/dropdown/components/SharedDropdownMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+import { MenuItemSelectAvatar } from '@/ui/navigation/menu-item/components/MenuItemSelectAvatar';
+import { getImageAbsoluteURIOrBase64 } from '~/utils/image/getImageAbsoluteURIOrBase64';
+
+export const SharedDropdownMenu = ({
+  messageThreadMembers,
+  label = '',
+  everyone = false,
+}: {
+  messageThreadMembers: MessageThreadMember[] | null;
+  label?: string;
```
[the remark this answers is not in the document]
Comment (pereira0x) [reply packages/twenty-front/src/modules/ui/layout/dropdown/components/SharedDropdownMenu.tsx:29]: We followed the same approach as with the other components, so we don't believe this is necessary.

File: packages/twenty-front/src/modules/ui/layout/dropdown/components/SharedDropdownMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+import { DropdownMenuItemsContainer } from '@/ui/layout/dropdown/components/DropdownMenuItemsContainer';
+import { DropdownMenuSeparator } from '@/ui/layout/dropdown/components/DropdownMenuSeparator';
+import { useDropdown } from '@/ui/layout/dropdown/hooks/useDropdown';
+import { MenuItem } from '@/ui/navigation/menu-item/components/MenuItem';
+import { MenuItemSelectAvatar } from '@/ui/navigation/menu-item/components/MenuItemSelectAvatar';
+import { getImageAbsoluteURIOrBase64 } from '~/utils/image/getImageAbsoluteURIOrBase64';
+
+export const SharedDropdownMenu = ({
+  messageThreadMembers,
+  label = '',
```
[the remark this answers is not in the document]
Comment (pereira0x) [reply packages/twenty-front/src/modules/ui/layout/dropdown/components/SharedDropdownMenu.tsx:25]: We already use Typescript that says that `MessageThreadMember` is `MessageThreadMember[]`.

File: packages/twenty-front/src/modules/activities/right-drawer/components/MessageThreadMembersBar.tsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+const StyledButtonContainer = styled.div`
+  align-items: center;
+  display: flex;
+  flex-direction: row;
+  margin-left: ${({ theme }) => theme.spacing(3)};
+`;
+
+export const MessageThreadMembersBar = () => {
+  const messageThread = useRecoilValue(messageThreadState);
```
[the remark this answers is not in the document]
Comment (pereira0x) [reply packages/twenty-front/src/modules/activities/right-drawer/components/MessageThreadMembersBar.tsx:15]: This is already handled.

**Comment by simaosanguinho** [comment]
Good afternoon everyone! Just pushed some changes, please tell us if we missed something.

**Comment by simaosanguinho** [comment]
Hi everyone. Regarding shared emails visibility change, me and @pereira0x  were told on Discord that we could not leverage useFindManyRecords to change the visibility on shared email threads, given that nested query filters are not supported yet. Any ideas how we can move forward now @charlesBochet?

**Comment by FelixMalfait** [comment]
Hey thanks for this and apologies for the slow review! We'll review and try to get it merged this week.
Do you think you can handle the merge conflicts? I know it's a pain sorry

**Comment by simaosanguinho** [comment]
No worries!! The merge conflicts will be solved by the end of tomorrow.

**Comment by pereira0x** [comment]
Hey, just letting you guys know that every conflict is resolved!

**Review by FelixMalfait — COMMENTED** [review]
Hey apologies it took months to review!!! That's great work you did here

File: packages/twenty-front/src/modules/activities/emails/components/EmailThreadMembersChip.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (FelixMalfait) [reply packages/twenty-front/src/modules/activities/emails/components/EmailThreadMembersChip.tsx:10]: I think we can drop everything related to "everyone" and focus on 1-to-1 sharing for v1. Maybe "everyone" should be managed through the permission layer instead. And here we'll manage something a bit more fine-grained with product implications which would be "thread subscribers" (people who would receive notifications when a new message is received for example)

**Review by lucasbordeau — APPROVED** [review]

**Pull request merged**